File: ml-service/agent_evolution/feedback.py
"""GEPA feedback adapter. Feedback is structured, bounded, and privacy checked."""

from dataclasses import dataclass
import hashlib
import re


@dataclass(frozen=True)
class GepaFeedback:
    score: float
    feedback: str
    objective_scores: dict[str, float]

    @property
    def content_hash(self) -> str:
        return hashlib.sha256(self.feedback.encode('utf-8')).hexdigest()


def build_feedback(*, candidate_id: str, score: float, objectives: dict[str, float], failures: list[str]) -> GepaFeedback:
    text = '\n'.join([
        f'Candidate: {candidate_id[:120]}',
        f'Result: {score:.3f}',
        'Feedback:',
        *[f'- {name}: {float(value):.3f}' for name, value in sorted(objectives.items())],
        *[f'- failure: {str(failure)[:300]}' for failure in failures[:12]],
    ])
    if re.search(r'email|phone|income|jwt|password|monthlyTakeHome|userId', text, re.I):
        raise ValueError('GEPA feedback contains private data')
    return GepaFeedback(score=float(score), feedback=text[:6000], objective_scores=dict(objectives))

File: server/agents/evolution/feedback.js
import { canonicalSha256 } from '../../utils/canonicalJson.js';

function safeText(value, max = 500) {
  return String(value || '').replace(/[\u0000-\u001f\u007f]/g, ' ').slice(0, max);
}

export function buildGepaFeedback({ candidateId, evaluation = null, reliability = null, authorityDelta = 0, failures = [], sandbox = null } = {}) {
  const cards = evaluation?.scoreCards || [];
  const passed = cards.filter(card => card.passed).length;
  const lines = [
    `Candidate: ${safeText(candidateId, 120)}`,
    `Result: ${cards.length ? (passed / cards.length).toFixed(3) : '0.000'}`,
    `Feedback:`,
    `- train/validation scorecards passed: ${passed}/${cards.length}`,
    `- financial authority delta: ${Number(authorityDelta) === 0 ? '0' : safeText(authorityDelta, 20)}`,
    `- reliability lab: ${reliability?.passed === true ? 'passed' : 'failed'}`,
    `- sandbox attestation: ${sandbox?.manifestHash ? 'present' : 'missing'}`,
  ];
  for (const card of cards.slice(0, 12)) {
    const failedGates = Object.entries(card.hardGates || {}).filter(([key, value]) => !['forbiddenTools', 'unsupportedTools', 'financialAuthorityDelta', 'authorityMeasurementState'].includes(key) && value === false).map(([key]) => key);
    if (failedGates.length) lines.push(`- ${card.partition} case failed gates: ${failedGates.join(', ')}`);
    if (card.scores?.actionCorrect === false) lines.push(`- ${card.partition} action mismatch`);
    if (card.scores?.grounded === false) lines.push(`- ${card.partition} grounding coverage failed`);
  }
  for (const failure of failures.slice(0, 12)) lines.push(`- failure: ${safeText(failure, 300)}`);
  return Object.freeze({ text: lines.join('\n'), contentHash: canonicalSha256(lines.join('\n')) });
}

export function assertGepaFeedbackSafe(feedback) {
  if (!feedback || typeof feedback.text !== 'string' || !feedback.contentHash) throw new Error('Structured GEPA feedback is required.');
  if (/email|phone|income|jwt|password|monthlyTakeHome|userId/i.test(feedback.text)) {
    const error = new Error('GEPA feedback contains sensitive data.');
    error.code = 'GEPA_FEEDBACK_PRIVACY_VIOLATION';
    throw error;
  }
  if (canonicalSha256(feedback.text) !== feedback.contentHash) throw new Error('GEPA feedback hash mismatch.');
  return true;
}
